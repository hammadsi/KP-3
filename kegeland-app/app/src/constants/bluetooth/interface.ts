import {DeviceStackParamList} from '~routes/interface';

export type BluetoothCharacteristic = string;
export type BluetoothService = {
  uuid: string;
  characteristics: BluetoothCharacteristic[];
};

export type BluetoothProfile = {
  navScreen: keyof DeviceStackParamList;
  services: string[];
  batteryService: {
    service: BluetoothService;
    fn: (bytes: number[]) => number;
  };
};

export type PeripheralNotification = {
  value: number[];
  peripheral: string;
  characteristic: string;
  service: string;
};
