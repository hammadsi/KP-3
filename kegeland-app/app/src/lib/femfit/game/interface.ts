import Matter from 'matter-js';
import {GameEngineUpdateEventOptionType} from 'react-native-game-engine';

export type Position = {
  x: number;
  y: number;
};

export type EntityBase = {
  body: Matter.Body;
};

export type ExerciseScheme = {
  name: string;
  description: string;
  repetitions: number;
  data: [number, number][];
};

export type GameEntities = {
  [key: string]: EntityBase & {[key: string]: any};
} & {
  physics: {
    engine: Matter.Engine;
    world: Matter.World;
  };
  exercise: ExerciseScheme & {currRep: number; currStep: number};
};

export type IGameEngineSystem = (
  entities: GameEntities,
  update: GameEngineUpdateEventOptionType,
) => GameEntities;
