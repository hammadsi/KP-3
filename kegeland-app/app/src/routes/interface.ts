import {
  CompositeScreenProps,
  NavigatorScreenParams,
} from '@react-navigation/native';
import type {BottomTabScreenProps} from '@react-navigation/bottom-tabs';
import type {StackScreenProps} from '@react-navigation/stack';

/**
 * Navigation interface for auth routes
 */
export type AuthStackParamList = {
  Login: undefined;
  Register: undefined;
  'Forgot password': undefined;
};

export type DeviceStackParamList = {
  Devices: undefined;
  Femfit: undefined;
};

/**
 * Navigation interface for settings routes
 */
export type SettingsStackParamList = {
  Settings: undefined;
  'Connect device': undefined;
};

/**
 * Base navigation interface
 */
export type RootTabParamList = {
  AuthStack: NavigatorScreenParams<AuthStackParamList> | undefined;
  DeviceStack: NavigatorScreenParams<DeviceStackParamList> | undefined;
  SettingsStack: NavigatorScreenParams<SettingsStackParamList> | undefined;
};

/**
 * Interface for screens on root-level
 */
export type RootScreenProps<
  RouteName extends keyof RootTabParamList = keyof RootTabParamList,
> = BottomTabScreenProps<RootTabParamList, RouteName>;

/**
 * Interface for screens under the auth-stack
 */
export type AuthScreenProps<
  RouteName extends keyof AuthStackParamList = keyof AuthStackParamList,
> = CompositeScreenProps<
  StackScreenProps<AuthStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

export type DeviceScreenProps<
  RouteName extends keyof DeviceStackParamList = keyof DeviceStackParamList,
> = CompositeScreenProps<
  StackScreenProps<DeviceStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

/**
 * Interface for screens under the settings-stack
 */
export type SettingsScreenProps<
  RouteName extends keyof SettingsStackParamList = keyof SettingsStackParamList,
> = CompositeScreenProps<
  StackScreenProps<SettingsStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;
