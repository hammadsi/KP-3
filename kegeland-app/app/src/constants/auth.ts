export enum UserRole {
  PATIENT = 'patient',
  PHYSICIAN = 'physician',
}

export enum Token {
  ACCESS_TOKEN = '@access_token',
  ID_TOKEN = '@id_token',
  REFRESH_TOKEN = '@refresh_token',
}

/**
 * How often to refresh the auth tokens
 */
export const REFRESH_INTERVAL_MS = 50 * 60 * 1000; // every 50th minute
