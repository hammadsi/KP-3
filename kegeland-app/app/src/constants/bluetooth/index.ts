import FEMFIT_PROFILE from '~lib/femfit/bluetooth/profile';

import {BluetoothProfile} from './interface';

/**
 * How long to scan for bluetooth peripherals, in seconds.
 */
export const SCAN_TIME = 5;
/**
 * Whether or not to allows duplicate devices
 */
export const ALLOW_DUPLICATE_DEVICES = false;
/**
 * How often to read BLE characteristic notifications
 */
export const UPDATE_INTERVAL_MS = 250;

export const _profiles = {
  femfit: FEMFIT_PROFILE,
} as const;

export type DeviceType = keyof typeof _profiles;

/**
 * Record of supported bluetooth profiles
 */
export const BLE_PROFILES = _profiles as Record<DeviceType, BluetoothProfile>;

/**
 * Record of supported peripherals
 */
export const PERIPHERAL_MAP: Record<string, DeviceType> = {
  [FEMFIT_PROFILE.batteryService.service.uuid]: 'femfit',
};
