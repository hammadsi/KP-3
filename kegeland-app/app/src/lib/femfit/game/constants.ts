import {Dimensions} from 'react-native';

/**
 * Constants for the femfit game
 */
export default {
  MAX_WIDTH: Dimensions.get('screen').width,
  MAX_HEIGHT: Dimensions.get('screen').height,
  BASELINE: Dimensions.get('screen').height * 0.6,
  COINS_PER_SEC: 1,
  HEADER_HEIGHT: 50,
  BOUNDARY_HEIGHT: 50,
  SPRITE_SIZE: 128,
  PLAYER_SIZE: 50,
  COIN_SIZE: 25,
  GRAVITY: 0.4,
};
