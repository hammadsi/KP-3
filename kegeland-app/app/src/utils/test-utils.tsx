/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
