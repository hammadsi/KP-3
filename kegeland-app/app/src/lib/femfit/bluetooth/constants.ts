import {BluetoothService} from '~constants/bluetooth/interface';
import {transformId} from '~utils/platform.utils';

export const PRESSURE_SENSOR_MIN_VAL = 760;
export const PRESSURE_SENSOR_MAX_VAL = 970;
export const ACTIVATION_THRESHOLD = 0.15;

/**
 * Transform UUIDs based on the current platform
 */
const HOUSEKEEPING_SERVICE_UUID = transformId(
  'b92a0001-4bf9-4870-8aa1-881b3a20ada4',
);
const HOUSEKEEPING_CHARACTERISTIC_UUID = transformId(
  'b92a0002-4bf9-4870-8aa1-881b3a20ada4',
);
const SENSOR_SERVICE_UUID = transformId('0d9e0001-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_1_UUID = transformId('0d9e0002-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_2_UUID = transformId('0d9e0003-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_3_UUID = transformId('0d9e0004-c111-49cd-bba3-85c7471cb6fa');
const SENSOR_4_UUID = transformId('0d9e0005-c111-49cd-bba3-85c7471cb6fa');

/**
 * The femfit services
 */
export const SERVICES = {
  HOUSEKEEPING_SERVICE: HOUSEKEEPING_SERVICE_UUID,
  SENSOR_SERVICE: SENSOR_SERVICE_UUID,
};

/**
 * The housekeeping service for femfit
 */
export const HOUSEKEEPING_SERVICE: BluetoothService = {
  uuid: HOUSEKEEPING_SERVICE_UUID,
  characteristics: [HOUSEKEEPING_CHARACTERISTIC_UUID],
};

/**
 * The sensor service for femfit
 */
export const SENSOR_SERVICE: BluetoothService = {
  uuid: SENSOR_SERVICE_UUID,
  characteristics: [SENSOR_1_UUID, SENSOR_2_UUID, SENSOR_3_UUID, SENSOR_4_UUID],
};
