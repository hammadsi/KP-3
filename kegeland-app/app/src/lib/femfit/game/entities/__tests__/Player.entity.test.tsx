import React from 'react';
import Matter from 'matter-js';

import {render} from '~utils/test-utils';

import Player, {PlayerRenderer, PlayerProps} from '../Player.entity';
import constants from '../../constants';
import {Position} from '../../interface';

const {PLAYER_SIZE} = constants;
describe('Test Player-entity', () => {
  const pos: Position = {x: 100, y: 100};
  it('should render correctly', () => {
    const props: PlayerProps = {
      body: Matter.Bodies.rectangle(pos.x, pos.y, PLAYER_SIZE, PLAYER_SIZE, {
        label: 'Player',
        restitution: 0,
        density: 0.0025,
        frictionAir: 0.015,
      }),
    };
    const component = <PlayerRenderer {...props} />;
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should spawn the entity correctly', () => {
    const world = Matter.World.create({});
    const entity = Player(world, pos);
    expect(Object.keys(entity).sort()).toStrictEqual(
      ['body', 'renderer'].sort(),
    );
  });
});
