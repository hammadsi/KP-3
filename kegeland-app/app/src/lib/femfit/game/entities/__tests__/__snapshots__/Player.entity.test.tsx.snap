// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Player-entity should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "overflow": "hidden",
        },
        Array [
          Object {
            "position": "absolute",
          },
          Object {
            "height": 50,
            "left": 75,
            "top": 75,
            "width": 50,
          },
        ],
      ]
    }
  >
    <Image
      collapsable={false}
      nativeID="animatedComponent"
      resizeMode="stretch"
      source={
        Object {
          "testUri": "../../../assets/femfit/bird.png",
        }
      }
      style={
        Object {
          "height": 128,
          "opacity": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": 384,
        }
      }
    />
  </View>
</View>
`;
